<!doctype html> <html lang=en > <meta charset=UTF-8 > <meta name=viewport  content="width=device-width, initial-scale=1"> <link rel=stylesheet  href="/libs/highlight/github.min.css"> <link rel=stylesheet  href="/css/franklin.css"> <link rel=stylesheet  href="/css/poole_lanyon.css"> <link rel=stylesheet  href="/css/adjust.css"> <link rel=stylesheet  href="/css/stackoverflow-light.css"> <link rel=icon  href="/assets/favicon.png"> <title>Think FASTA</title> <input type=checkbox  class=sidebar-checkbox  id=sidebar-checkbox > <div class=sidebar  id=sidebar > <div class=sidebar-item > <p>A theme adapted from <a href="http://lanyon.getpoole.com/" target=_blank >Lanyon</a>.</p> </div> <nav class=sidebar-nav > <a class="sidebar-nav-item " href="/">Home</a> <a class="sidebar-nav-item " href="/syllabus/">Syllabus</a> <a class="sidebar-nav-item " href="/lessons/">Lessons</a> <a class="sidebar-nav-item " href="/lectures-labs/">Lectures & Labs</a> <a class="sidebar-nav-item " href="/assignments/">Assignments</a> </nav> <div class=sidebar-item > <p>&copy; Kevin Bonham, PhD.</p> </div> </div> <!-- Wrap is the content to shift when toggling the sidebar. We wrap the content to avoid any CSS collisions with our real content. --> <div class=wrap > <div class=masthead > <div class=container > <h3 class=masthead-title > <a href="/" title=Home >Wellesley College - BISC189</a> <small>Essential Skills for Computational Biology</small> </h3> </div> </div> <div class="container content"> <div class=franklin-content > <h1 id=a_hrefassignmentsassignment06assignment_6_-_think_fasta ><a href="#a_hrefassignmentsassignment06assignment_6_-_think_fasta" class=header-anchor ><a href="/assignments/Assignment06">Assignment 6 - Think FASTA</a></a></h1> <div class=badges ><a href="https://classroom.github.com/a/VRVmT_th"><img src="https://img.shields.io/badge/Github&#37;20-6-darkblue?style&#61;for-the-badge" alt=assignment6link  /></a> <img src="https://img.shields.io/badge/Due-2021--06--28-red?style&#61;for-the-badge" alt=assignment6due  /></div> <h2 id=instructions ><a href="#instructions" class=header-anchor >Instructions</a></h2> <p>Click the assignment 6 invitation above, clone the repository, and follow the instructions in <code>src/assignment.jl</code>.</p> <p>Also, recall that you may run the automated tests on your own computer:</p> <pre><code class=language-sh >&#36; julia --project -e &#39;using Pkg; Pkg.test&#40;&#41;&#39;</code></pre>
<h2 id=assignment06_code ><a href="#assignment06_code" class=header-anchor >Assignment06 code</a></h2>
<p>For each assignment, the contents of the assignment code script will be rendered as html at the bottom of the assignment description page. If you&#39;re interested in how that works, check out <a href="https://fredrikekre.github.io/Literate.jl/v2/">Literate.jl</a></p>
<h1 id=instructions_for_assignment06 ><a href="#instructions_for_assignment06" class=header-anchor >Instructions for Assignment06</a></h1>
<h2 id=introduction ><a href="#introduction" class=header-anchor >Introduction</a></h2>
<p>Note: this file is runnable in its current state, but is incomplete. You can run the file from the command line, or use the VS Code julia extension to run individual lines.</p>
<h2 id=question_1_-_reading_fasta_files ><a href="#question_1_-_reading_fasta_files" class=header-anchor >Question 1 - Reading FASTA files</a></h2>
<p>A common way to interact with biological sequence data is in a format called <a href="https://en.wikipedia.org/wiki/FASTA_format"><code>FASTA</code></a>. This is a very simple format, where each &quot;record&quot; &#40;sequence&#41; contains:</p>
<ol>
<li><p>A &quot;header&quot; line, always starting with <code>&gt;</code>, that contains a description of the sequence</p>

<li><p>The sequence, which can be split across multiple lines</p>

</ol>
<p>For example, in the <code>data/</code> directory of this assignment, there is a file called <code>cov2_genomes.fasta</code>, which contains some genome sequences of Sars-CoV2 isolated from humans early in the pandemic. Use the shell to take a look at the first 10 lines &#40;<code>head</code>&#41; of the file.</p>
<p>Now, let&#39;s work on building a julia function to help us use this file.</p>
<pre><code class=language-julia >&quot;&quot;&quot;
    count_records&#40;path&#41;

Given a path to a &#96;fasta&#96; file,
counts and returns the number of records it contains.

Example
≡≡≡≡≡≡≡≡≡
    julia&gt; count_records&#40;&quot;data/cov2_genomes.fasta&quot;&#41;
    10
&quot;&quot;&quot;
function count_records&#40;path&#41;
    # Your code here.
end</code></pre>
<p>Hint - If you&#39;re having trouble, the easiest thing to do is probably to count lines that start with <code>&gt;</code>.</p>
<h2 id=question_2_-_split_the_headers ><a href="#question_2_-_split_the_headers" class=header-anchor >Question 2 - Split the headers</a></h2>
<p>It&#39;s nice to know the number of records the file contains, but it&#39;s more useful to actually be able to read and store those records.</p>
<p>Let&#39;s look again at the headers for these sequences. This can be accomplished from the terminal using <code>grep</code>. The simplest way to use <code>grep</code> is with 2 arguments - the first argument specifies a pattern to look for, and the second is the path to a file &#40;or group of files&#41;.</p>
<p><code>grep</code> will then print to the screen all of the lines from the file that match that pattern. To see all of the header lines, for example,</p>
<pre><code class=language-sh >&#36; grep &#39;&gt;&#39; data/cov2_genomes.fasta</code></pre>
<p>Each header has 3 components, separated by <code>|</code> Write a function that divides the header into these 3 components, and returns them as a tuple.</p>
<pre><code class=language-julia >&quot;&quot;&quot;
    fasta_header&#40;header&#41;

Divides a fasta header into its component parts,
removing any leading or trailing spaces.

Example
≡≡≡≡≡≡≡≡≡
    julia&gt; fasta_header&#40;&quot;&gt;M0002 |China|Homo sapiens&quot;&#41;
    &#40;&quot;M0002&quot;, &quot;China&quot;, &quot;Homo sapiens&quot;&#41;

    julia&gt; fasta_header&#40;&quot;AAATTC&quot;&#41;
    Error: Invalid header &#40;headers must start with &#39;&gt;&#39;&#41;

    julia&gt; fasta_header&#40;&quot;&gt;Another sequence&quot;&#41;
    &#40;&quot;Another sequence&quot;,&#41;

    julia&gt; fasta_header&#40;&quot;&gt;headers| can | have| any number | of | info blocks&quot;&#41;
    &#40;&quot;headers&quot;, &quot;can&quot;, &quot;have&quot;, &quot;any number&quot;, &quot;of&quot;, &quot;info blocks&quot;&#41;
&quot;&quot;&quot;
function fasta_header&#40;header&#41;
    startswith&#40;header, &#39;&gt;&#39;&#41; || error&#40;&quot;Invalid header &#40;headers must start with &#39;&gt;&#39;&#41;&quot;&#41;

    # Your code here
end</code></pre>
<p>A couple of things to note about this function if you&#39;re strugling:</p>
<ol>
<li><p>Notice that, even though each header must start with a <code>&gt;</code>, the first item in the returned Tuple should not</p>

<li><p>The return item should always be a Tuple, even if there&#39;s only 1 item</p>

<li><p>The <code>split&#40;&#41;</code> function can divide a <code>String</code> by a delimeter.</p>

<li><p>You may need to do some searching on the internet.</p>

</ol>
<h2 id=question_3_-_parse_the_file ><a href="#question_3_-_parse_the_file" class=header-anchor >Question 3 - parse the file</a></h2>
<p>Alright - let&#39;s actually parse the file. For now, the easiest thing to do is to return two vectors, one containing the parsed identifiers as tuples, and one containing the sequences as strings.</p>
<p>One thing to keep in mind is that the sequences in a fasta file can be spread out over multiple lines, but they&#39;ll be much easier to work with as a single string, so your function should have a way to join multiple lines &#40;remember the <code>join&#40;&#41;</code> function?&#41;.</p>
<p>I&#39;ve included a few example files in the &quot;data&quot; directory with some example files to test your functions on. The examples in the docstring below assume that your working directory is the assignment directory.</p>
<pre><code class=language-julia >&quot;&quot;&quot;
    function parse_fasta&#40;path&#41;

Reads a fasta-formated file and returns 2 vectors,
one containing parsed headers,
the other containing the entire sequence as a &#96;String&#96;.

Note: function does not validate DNA sequences for correctness.

Example
≡≡≡≡≡≡≡≡≡
    julia&gt; ex1 &#61; parse_fasta&#40;&quot;data/ex1.fasta&quot;&#41;;

    julia&gt; ex1 isa Tuple
    true

    julia&gt; ex1&#91;1&#93;
    2-element Array&#123;Tuple&#123;String,String&#125;,1&#125;:
      &#40;&quot;ex1.1&quot;, &quot;easy&quot;&#41;
      &#40;&quot;ex1.2&quot;, &quot;multiline&quot;&#41;

    julia&gt; ex1&#91;2&#93;
    2-element Array&#123;String,1&#125;:
    &quot;AATTATAGC&quot;
    &quot;CGCCCCCCAGTCGGATT&quot;

    julia&gt; ex2 &#61; parse_fasta&#40;&quot;data/ex2.fasta&quot;&#41;;

    julia&gt; ex2&#91;1&#93;
    4-element Array&#123;Tuple&#123;String,String&#125;,1&#125;:
      &#40;&quot;ex2.1&quot;, &quot;oneper&quot;&#41;
      &#40;&quot;ex2.2&quot;, &quot;wrong&quot;&#41;
      &#40;&quot;ex2.3&quot;, &quot;wronger&quot;&#41;
      &#40;&quot;ex2.4&quot;, &quot;wrongest&quot;&#41;

    julia&gt; ex2&#91;2&#93;
    4-element Array&#123;String,1&#125;:
      &quot;ATCCGT&quot;
      &quot;ATCGTGGaact&quot;
      &quot;ATCGTGGaact&quot;
      &quot;this isn&#39;t a dna string,but parse it anyway&quot;
&quot;&quot;&quot;
function parse_fasta&#40;path&#41;
    # Think through the components you need
    # Does it make sense to define any containers at the beginning?
    # How will you loop through the file?
    # What do you need to get from each line?
end</code></pre>
<p>Tips:</p>
<ol>
<li><p>Build the function in parts</p>
<ul>
<li><p>first, see if you can build the vector of headers, ignoring the sequences</p>

<li><p>next, see if you can collect the sequences, even if they&#39;re not joined together properly</p>

</ul>

<li><p>Work on pieces of the function in the REPL. Make sure each expression does what you think it does.</p>
<p>For example, if you make a <code>for</code> loop to go through each line of the file, add <code>@info</code> statements to check the value of each variable in each loop.</p>

</ol>
<h2 id=question_4_-_sequence_statistics ><a href="#question_4_-_sequence_statistics" class=header-anchor >Question 4 - Sequence Statistics</a></h2>
<p>We now have most of the pieces necessary to get some summary statistics of CoV2 genomes. Note: You may want to grab some code from previous assignments.</p>
<p>Calculate the mean sequence length and standard deviation of the CoV2 geneomes in <code>data/cov2_genomes.fasta</code>.</p>
<pre><code class=language-julia ># helper code goes here

mean_cov2_length &#61; 0.0
std_cov2_length &#61; 0.0</code></pre>
<p>Calculate the mean and standard deviation of GC content of the CoV2 genomes in <code>data/cov2_genomes.fasta</code>. GC content should be a number between 0 and 1. Note: if you already parsed the fasta file above and assigned the results to variables, there&#39;s no need to re-parse it.</p>
<pre><code class=language-julia ># helper code goes here

mean_cov2_gc &#61; 0.0
std_cov2_gc &#61; 0.0</code></pre>
<h2 id=bonus_question ><a href="#bonus_question" class=header-anchor >Bonus Question</a></h2>
<p>Write a function that will read any proper fasta file and returns a tuple with the mean and standard deviation of the sequence length and gc content.</p>


<div class=page-foot >
  <div class=copyright >
    &copy; Kevin Bonham, PhD. Last modified: August 18, 2021. Website built with <a href="https://github.com/tlienart/Franklin.jl">Franklin.jl</a> and the <a href="https://julialang.org">Julia programming language</a>.
  </div>
</div>
</div>
      </div>  
    </div> 
    
    
        <script src="/libs/highlight/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();hljs.configure({tabReplace: '    '});</script>

    
    <label for=sidebar-checkbox  class=sidebar-toggle ></label>